File: problems-adapt/0201-0300/0260_two-colorable-graph/solution_union_find.ts
function isTwoColorable(graph: number[][]): boolean {
    const n = graph.length;
    const parent: number[] = Array.from({ length: n }, (_, i) => i);

    const find = function (node: number): number {
        let root = node;
        while (parent[root] !== root) {
            root = parent[root];
        }
        // Second walk repoints every visited node at the root (path
        // compression), flattening the structure for later finds.
        while (parent[node] !== root) {
            const next = parent[node];
            parent[node] = root;
            node = next;
        }
        return root;
    };

    const union = function (a: number, b: number): void {
        parent[find(a)] = find(b);
    };

    // Two-colorable means the nodes split into two groups with every edge
    // crossing between them, so all of a node's neighbors must be able
    // to share the one opposite group.
    for (let u = 0; u < n; u++) {
        for (const v of graph[u].slice(1)) {
            // Union u's enemies together: they all belong to one set.
            union(graph[u][0], v);
        }
    }
    // A node sharing a set with one of its own enemies sits inside an
    // odd cycle: not two-colorable.
    for (let u = 0; u < n; u++) {
        for (const v of graph[u]) {
            if (find(u) === find(v)) {
                return false;
            }
        }
    }
    return true;
}

File: problems/0701-0800/0785_two-colorable-graph/solution_union_find.cpp
class Solution {
    vector<int> parent;

  public:
    bool isTwoColorable(vector<vector<int>> &graph) {
        int n = graph.size();
        parent.assign(n, 0);
        iota(parent.begin(), parent.end(), 0);
        // Two-colorable means the nodes split into two groups with every edge
        // crossing between them, so all of a node's neighbors must be able
        // to share the one opposite group.
        for (int u = 0; u < n; u++) {
            for (size_t v = 1; v < graph[u].size(); v++) {
                // Union u's enemies together: they all belong to one set.
                myUnion(graph[u][0], graph[u][v]);
            }
        }
        // A node sharing a set with one of its own enemies sits inside an
        // odd cycle: not two-colorable.
        for (int u = 0; u < n; u++) {
            for (int v : graph[u]) {
                if (find(u) == find(v)) {
                    return false;
                }
            }
        }
        return true;
    }

  private:
    int find(int node) {
        int root = node;
        while (parent[root] != root) {
            root = parent[root];
        }
        // Second walk repoints every visited node at the root (path
        // compression), flattening the structure for later finds.
        while (parent[node] != root) {
            int next = parent[node];
            parent[node] = root;
            node = next;
        }
        return root;
    }

    void myUnion(int a, int b) { parent[find(a)] = find(b); }
};
